src/components/PDFView/index.native.js

```diff
@@ -24,24 +25,123 @@ const defaultProps = {
 };
 
 /**
- * On the native layer, we use a pdf library to display PDFs
+ * On the native layer, we use a pdf library to display PDFs. If a PDF is
+ * password-protected we render a PDFPasswordForm to request a password
+ * from the user.
  *
- * @param props
- * @returns {JSX.Element}
+ * In order to render things nicely during a password challenge we need
+ * to keep track of a bunch of additional state. In particular, the
+ * react-native-pdf/PDF component is both conditionally rendered and hidden
+ * depending upon the situation. It needs to be rerendered on each password
+ * submission because it doesn't dynamically handle updates to it's
+ * password property. And we need to hide it during password challenges
+ * so that PDFPasswordForm doesn't bounce when react-native-pdf/PDF
+ * is (temporarily) rendered.
  */
+class PDFView extends PureComponent {
+    constructor(props) {
+        super(props);
+        this.state = {
+            shouldRequestPassword: false,
+            isPasswordInvalid: false,
+            password: '',
+            shouldAttemptPdfLoad: true,
+            isPdfLoaded: false,
+        };
+        this.initiatePasswordChallenge = this.initiatePasswordChallenge.bind(this);
+        this.attemptPdfLoadWithPassword = this.attemptPdfLoadWithPassword.bind(this);
+        this.terminatePasswordChallenge = this.terminatePasswordChallenge.bind(this);
+    }
 
-const PDFView = props => (
-    <TouchableWithoutFeedback style={[styles.flex1, props.style]}>
-        <PDF
-            activityIndicator={<FullScreenLoadingIndicator />}
-            source={{uri: props.sourceURL}}
-            style={[
-                styles.imageModalPDF,
-                StyleUtils.getWidthAndHeightStyle(props.windowWidth, props.windowHeight),
-            ]}
-        />
-    </TouchableWithoutFeedback>
-);
+    /**
+     * Initiate password challenge if message received from react-native-pdf/PDF
+     * indicates that a password is required is invalid.
+     *
+     * For a password challenge the message is "Password required or incorrect password."
+     * Note that the message doesn't specify whether the password is simply empty or
+     * rather invalid.
+     *
+     * @param {String} message
+     */
+    initiatePasswordChallenge({message}) {
```

Comment:
Yes create a new BasePDFView.js

Reply:
Ok thanks, @Santhosh-Sellavel.

And sorry, I do see examples like this.

Reply:
Ok sounds good.

Reply:
Hey so sorry, I don't see a nice way to do this. If I understand correctly we should not extend components. So doing something like in `BaseTextInput.js`, for example, doesn't seem possible to me.

The main issue is that things need to be handled quite differently on web and on native in order to acheive a proper password challenge experience:

1. On web, the `react-pdf` component **must not** be conditionally rendered - and thus hidden via styles when the password form is visible. This is because it receives the password via a callback function and so the PDF component needs to be persistent.

2. Whereas on native, the `react-native-pdf` **must** be conditionally rendered. This is because it doesn't detect when its `props.password` is updated.

Both versions use similar state variables (`shouldRequestPassword` and `isPasswordInvalid`) but otherwise the code totally different.

Can someone maybe provide some specific direction on how to refactor this?

Reply:
OK, thank you for that explanation. I agree with you now that this is probably fine the way it is.